### PR TITLE
Improving feature names

### PR DESCRIPTION
**Pull request by jbschiratti**
This PR aims at having more meaningful feature names when `extract_features`is called with `return_df = True`.

### CODE REVIEW

File: mne_features/feature_extraction.py [hunk 1]
```diff
@@ -67,12 +68,40 @@ def transform(self, X, y='deprecated'):
...
+                            range(self.n_channels) for _, i, j in
+                            _idxiter(n_freq_bands, triu=False)]
+            pow_names = ['ch%s_%s' % (ch_num, i) for ch_num in
+                         range(self.n_channels) for i in range(n_freq_bands)]
+            if _params['ratios'] is None:
+                return pow_names
+            elif _params['ratios'] == 'only':
+                return ratios_names
+            else:
+                return pow_names + ratios_names
```
Comment (agramfort) [inline mne_features/feature_extraction.py]: this is not really clean. This gymnastic should be done via the function pow_freq_bands. You have some logic in a function agnostic class which percolates from a custom function. You should attach this logic to the pow_freq_bands callable.

Comment (jbschiratti) [reply mne_features/feature_extraction.py]: I agree, this is not clean. This first commit was just to allow @l-omar-chehab  to move on (and have meaning feature names for `compute_pow_freq_bands`). Now, we can think about making it nice.

**Comment by jbschiratti** [comment]
Here is one idea on how the feature names could be "attached" to the feature functions.

If called with `ratios=None`, the feature function `compute_pow_freq_bands` will return an array with shape `(n_channels * n_freq_bands,)`. In this case, possible [meaningful] feature names could be `ch0_band0`, `ch0_band1`, `ch0_band2`..., `ch1_band0`,... To "attach" this information to `compute_pow_freq_bands` we could use a decorator:
```
@with_feature_names('ch[0]_band[1]')
def compute_pow_freq_bands(sfreq, data, ...)
    ...
```
So that, in `FeatureFunctionTransformer`, `self.func.feat_names_pattern` gives: `'ch[0]_band[1]`. Then, in the `get_feature_names` method of `FeatureFunctionTransformer`, we could have something like this:
```
pattern = self.func.feat_names_pattern  # 'ch[0]_band[1]'
feature_names = self._get_feature_names_helper(pattern, self.out_shape)
return feature_names
```
where `out_shape` corresponds to `X_out.shape` with `X_out`, the array returned by the feature function. For this to work, feature functions will need to return multidimensional ndarrays. This is not a big issue since the `transform` method of `FeatureFunctionTransformer` could be changed to return `X_out.ravel()` instead of `X_out`.

In the code above, `_get_feature_names_helper` would transform `'ch[0]_band[1]'` into the list `['ch0_band0', 'ch0_band1', ..., 'ch1_band0', 'ch1_band1',...]` with using the following rationale: `[0]` in the input string mean iterating from `0` to `out_shape[0] - 1` and `[1]` iterating from `0` to `out_shape[1] - 1`. This way, `'ch[0]_band[1]` would be equivalent to: 

```
['ch%s_band%s' % (i, j) for i in range(out_shape[0]) for j in range(out_shape[1])]
```

... you get the idea!

If a feature function does not have a `feat_names_pattern` attribute, nothing changes.

File: mne_features/feature_extraction.py [hunk 2]
```diff
@@ -66,13 +67,19 @@ def transform(self, X, y='deprecated'):
             details.
         """
         X_out = super(FeatureFunctionTransformer, self).transform(X, y)
+        _feature_func = _get_python_func(self.func)
+        _params = self.get_params()
+        if hasattr(_feature_func, 'get_feature_names'):
+            self.feature_names = _feature_func.get_feature_names(X, **_params)
```
Comment (agramfort) [inline mne_features/feature_extraction.py]: I don't like that calling transform affects the state of the object. Only a fit is allowed to do this. Can you see a way out? also any attribute that is data dependent should end with _

Comment (jbschiratti) [reply mne_features/feature_extraction.py]: I see. Then, what about
```
    def fit(self, X, y=None):
        """Fit the FeatureFunctionTransformer (does not extract features).
        
        Parameters
        ----------
        X : ndarray, shape (n_channels, n_times)
        
        y : ignored
        
        Returns
        -------
        self
        """
        self._check_input(X)
        _feature_func = _get_python_func(self.func)
        _params = self.get_params()
        if hasattr(_feature_func, 'get_feature_names'):
            self.feature_names = _feature_func.get_feature_names(X, **_params)
        return self
```
in `FeatureFunctionTransformer`?

Comment (agramfort) [reply mne_features/feature_extraction.py]: yes this is ok in terms of API but `_params = self.get_params()` could be in the if block

**Comment by jbschiratti** [comment]
Test added !

File: mne_features/tests/test_univariate.py [hunk 3]
```diff
@@ -178,6 +179,40 @@ def test_pow_freq_bands():
                                                ratios='only'), expected_ratios)
 
 
+def test_feature_names_pow_freq_bands():
+    _data = data[:, :3, :]  # keep only 3 channels for the sake of simplicity
+    selected_funcs = ['pow_freq_bands']
+    fb = np.array([[4., 8.], [30., 70.]])
+    ratios_col_names = ['ch0_0_1', 'ch0_1_0', 'ch1_0_1', 'ch1_1_0',
+                        'ch2_0_1', 'ch2_1_0']
+    pow_col_names = ['ch0_0', 'ch0_1', 'ch1_0', 'ch1_1', 'ch2_0', 'ch2_1']
```
Comment (agramfort) [inline mne_features/tests/test_univariate.py]: no way to have more explicit names likes alpha, beta etc?

Comment (jbschiratti) [reply mne_features/tests/test_univariate.py]: We could improve that but, at some point, the user would need to name the frequency bands he wishes to use. What about allowing the `freq_bands` parameter in `compute_pow_freq_bands` to be a dict as the one below? 
```
freq_bands = {'delta': [0.5, 4], 
              'theta': [4, 8], 
              'alpha': [8, 13], 
              'beta': [13, 30], 
              'low-gamma': [30, 70], 
              'high-gamma': [70, 100]}
```

**Comment by agramfort** [comment]
yes +1 !

**Comment by jbschiratti** [comment]
The last commit improves the feature names when `freq_bands` is a dict (such as the one above).

**Comment by agramfort** [comment]
Thanks

**Pull request merged**